File: hydra-haskell/src/main/haskell/Hydra/Dsl/Meta/Lib/Equality.hs
-- | Phantom-typed term DSL for the hydra.lib.equality library

{-# LANGUAGE FlexibleContexts #-}

module Hydra.Dsl.Meta.Lib.Equality where

import Hydra.Dsl.AsTerm
import Hydra.Dsl.Meta.Phantoms
import Hydra.Phantoms
import Hydra.Util
import qualified Hydra.Dsl.Terms as Terms
import Hydra.Sources.Libraries

import Data.Int


compare :: AsTerm t2 a => TTerm a -> t2 -> TTerm Comparison
compare t1 t2 = primitive2 _equality_compare t1 (asTerm t2)

equal :: AsTerm t2 a => TTerm a -> t2 -> TTerm Bool
equal t1 t2 = primitive2 _equality_equal t1 (asTerm t2)

gt :: AsTerm t2 a => TTerm a -> t2 -> TTerm Bool
gt t1 t2 = primitive2 _equality_gt t1 (asTerm t2)

gte :: AsTerm t2 a => TTerm a -> t2 -> TTerm Bool
gte t1 t2 = primitive2 _equality_gte t1 (asTerm t2)

identity :: TTerm a -> TTerm a
identity = primitive1 _equality_identity

lt :: AsTerm t2 a => TTerm a -> t2 -> TTerm Bool
lt t1 t2 = primitive2 _equality_lt t1 (asTerm t2)

lte :: AsTerm t2 a => TTerm a -> t2 -> TTerm Bool
lte t1 t2 = primitive2 _equality_lte t1 (asTerm t2)

max :: AsTerm t2 a => TTerm a -> t2 -> TTerm a
max t1 t2 = primitive2 _equality_max t1 (asTerm t2)

min :: AsTerm t2 a => TTerm a -> t2 -> TTerm a
min t1 t2 = primitive2 _equality_min t1 (asTerm t2)

File: hydra-java/src/main/java/hydra/dsl/meta/lib/Equality.java
package hydra.dsl.meta.lib;

import hydra.core.Name;
import hydra.dsl.meta.Expr;
import hydra.phantoms.TTerm;

import static hydra.dsl.meta.Phantoms.*;

/**
 * Phantom-typed term DSL for the hydra.lib.equality library.
 *
 * <p>Mirrors {@code Hydra.Dsl.Meta.Lib.Equality} (Haskell) and
 * {@code hydra.dsl.meta.lib.equality} (Python).
 *
 * <p>Usage: {@code import static hydra.dsl.meta.lib.Equality.*;}
 */
public interface Equality {

    Name _equality_compare = new Name("hydra.lib.equality.compare");
    Name _equality_equal = new Name("hydra.lib.equality.equal");
    Name _equality_gt = new Name("hydra.lib.equality.gt");
    Name _equality_gte = new Name("hydra.lib.equality.gte");
    Name _equality_identity = new Name("hydra.lib.equality.identity");
    Name _equality_lt = new Name("hydra.lib.equality.lt");
    Name _equality_lte = new Name("hydra.lib.equality.lte");
    Name _equality_max = new Name("hydra.lib.equality.max");
    Name _equality_min = new Name("hydra.lib.equality.min");

    static <R> Expr<R> compare(TTerm<?> x, TTerm<?> y) {
        return primitive2(_equality_compare, x, y);
    }

    static <R> Expr<R> equal(TTerm<?> a, TTerm<?> b) {
        return primitive2(_equality_equal, a, b);
    }

    static <R> Expr<R> gt(TTerm<?> a, TTerm<?> b) {
        return primitive2(_equality_gt, a, b);
    }

    static <R> Expr<R> gte(TTerm<?> a, TTerm<?> b) {
        return primitive2(_equality_gte, a, b);
    }

    static <R> Expr<R> identity(TTerm<?> a) {
        return primitive1(_equality_identity, a);
    }

    static <R> Expr<R> lt(TTerm<?> a, TTerm<?> b) {
        return primitive2(_equality_lt, a, b);
    }

    static <R> Expr<R> lte(TTerm<?> a, TTerm<?> b) {
        return primitive2(_equality_lte, a, b);
    }

    static <R> Expr<R> max(TTerm<?> a, TTerm<?> b) {
        return primitive2(_equality_max, a, b);
    }

    static <R> Expr<R> min(TTerm<?> a, TTerm<?> b) {
        return primitive2(_equality_min, a, b);
    }
}
